### ISSUE TITLE
t.w.twcgi does not allow multiple headers

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4742]
|<img alt="teolicy's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| teolicy reported|
|-|-|
|Trac ID|trac#4742|
|Type|defect|
|Created|2010-11-22 16:34:53Z|
|Branch|https://github.com/twisted/twisted/tree/cgi-repeated-header-4742|

t.w.twcgi uses request.setHeader instead of request.responseHeaders.addRawHeader. 

The former causes headers that appear more than once in the CGI script's output to be overwritten, violating HTTP RFC2616.

The included patch adds a test and a fix, the fix makes t.w.twcgi behave more like t.w.wsgi in this regard (use request.responseHeaders.addRawHeader, don't allow overwriting of the 'server' and 'date' headers).

Attachments:

* [t.w.twcgi-setHeader.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d28/d289f071936e9abc61411f4cc5232e4d964277b5/c926fd24461e56e32eb45cf1cfe4c46568f96d04.patch) (2271 bytes) - added by teolicy on 2010-11-22 16:45:36Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4742 4742
type__defect defect
reporter__teolicy teolicy
priority__normal normal
milestone__ 
branch__branches_cgi_repeated_header_4742 branches/cgi-repeated-header-4742
branch_author__teolicy teolicy
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1290443693000000 1290443693000000
changetime__1290483956000000 1290483956000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [30334]) Branching to 'cgi-repeated-header-4742'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [30335]) Apply t.w.twcgi-setHeader.patch 

refs [#4742](https://github.com/twisted/twisted/issues/4742)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [30336]) Update testing style:

  * test methods are prefixed with `test_` now
  * the old `_testCGI_1` is kept close to the method that uses it
  * the new `_testCGI_2` is replaced with a nested function in the test method that uses it
  * `assertEquals` is preferred over `failUnlessEqual`
  * the test method gets a docstring

refs [#4742](https://github.com/twisted/twisted/issues/4742)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to teolicy|
|-|-|

Thanks.  I checked the patch into a branch and made some cleanups to the test.  It looks like there's no test coverage for the server/date header checks though.  Would you like to add a test for that?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [30337]) Add a test for the server/date header protection

refs [#4742](https://github.com/twisted/twisted/issues/4742)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

I added a test for the server/date thing.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @exarkun|
|-|-|

There should be a separate ticket to somehow inform CGI authors that they're not allowed to manipulate 'Server' and 'Date'.

And probably you shouldn't be able to set multiple content-length headers, but maybe there's enforcement code for that elsewhere?

Generally though, this looks great.  Please land.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [30350]) Merge cgi-repeated-header-4742 

Author: teolicy, exarkun
Reviewer: exarkun, glyph
Fixes: [#4742](https://github.com/twisted/twisted/issues/4742)

Change `twisted.web.twcgi` to append headers to the response, rather than
replacing them, so that when a CGI emits multiple instances of a single
header (eg, two Set-Cookie headers), they are all sent in the response
to the client.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
